### ISSUE TITLE
The implementation in source code of Backpropagation in TensorFlow (Conv2DBackpropFilter and Conv2DBackpropInput)

### ISSUE THREAD

**Issue by Jiawen1991** [tensorflow/tensorflow#14109]
Hi,

Since two operations Conv2DBackpropFilter and Conv2DBackpropInput count most of the time for lots of applications(AlexNet/VGG/GAN/Inception, etc.), I am analyzing the complexity of these two operations (back-propagation) in TensorFlow and I found out that there are three implementation versions (custom, fast and slot) for Conv2DBackpropFilter (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_grad_filter_ops.cc ) and Conv2DBackpropInput (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_grad_input_ops.cc). While I profile, all computations are passed to "custom" version instead of "fast" or "slow" which directly calls Eigen function SpatialConvolutionBackwardInput to do that. 

The issue is:
Conv2DBackpropFilter uses Eigen:“TensorMap.contract" to do the tensor contraction and Conv2DBackpropInput uses Eigen:"MatrixMap.transpose" to do the matrix transposition in the Compute() function. Beside these two functions, I didn't see any convolutional operations which are needed for back-propagation theoretically. Beside convolutions, what else would be run inside these two operations for back-propagation? Does anyone know how to analyze the computation complexity of "back propagation" operation in TensorFlow?

I am looking for any advise/suggestion. Thank you!

**Comment by asimshankar** [comment]
I

**Comment by asimshankar** [comment]
I don't think I quite follow your question. As you pointed out, there are a few different implementations of those kernels, and the `REGISTER_KERNEL_BUILDER` calls in the files you listed will tell you which conditions a particular kernel implementation is applicable for.

The `Compute()` methods of these classes lists out exactly the code run when these operations are invoked.

Could you be more detailed on what you're looking for?
(Also, perhaps this question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. )

**Comment by Jiawen1991** [comment]
Asim, thank you for your reply. I asked in StackOverflow but no one responses me yet and I am in a hurry in this issue.

To be simple, the issue is that the normal Back-Propagation should include convolution, differentiate,  derivate and other operations to calculate a gradient. However, in TensorFlow, it only uses tensor contraction (for Conv2DBackpropFilter) and matrix transposition(for Conv2DBackpropInput) to do the Back-Propagation. How can it be?

**Comment by asimshankar** [comment]
I'm still not sure I understand the question, but to be clear, if you want to see all the operations conducted during the backpropagation pass then you should inspect the graph. The gradient computation is added to the graph by `tf.gradients`. You can use TensorBoard's [graph visualization](https://www.tensorflow.org/get_started/graph_viz) to see all the computation done in the forward and backward pass.

Hope that helps.

**Comment by Jiawen1991** [comment]
Thank you though I still can not find the operations workflow of backpropagation in TensorBoard's generated graph.

**Comment by angersson** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. Please move your issue there. Thanks!
